File: backend/src/main/java/com/samsung/hr/service/InterviewService.java
package com.samsung.hr.service;

import com.samsung.hr.domain.InterviewSession;
import com.samsung.hr.domain.ScoreTurn;
import com.samsung.hr.dto.*;
import com.samsung.hr.repository.InterviewSessionRepository;
import com.samsung.hr.repository.ScoreTurnRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.UUID;

@Service
public class InterviewService {

    private final InterviewSessionRepository sessionRepository;
    private final ScoreTurnRepository turnRepository;
    private final TemporalInterviewService temporalService;

    public InterviewService(
            InterviewSessionRepository sessionRepository,
            ScoreTurnRepository turnRepository,
            TemporalInterviewService temporalService) {
        this.sessionRepository = sessionRepository;
        this.turnRepository = turnRepository;
        this.temporalService = temporalService;
    }

    @Transactional
    public SessionResponse createSession(CreateSessionRequest request) {
        InterviewSession session = new InterviewSession();
        session.setInterviewerName(request.interviewerName());
        session.setCandidateName(request.candidateName());
        session.setRoleTitle(request.roleTitle());
        session = sessionRepository.save(session);
        return toSessionResponse(session);
    }

    public SessionResponse getSession(UUID id) {
        return sessionRepository
                .findById(id)
                .map(this::toSessionResponse)
                .orElseThrow(() -> new IllegalArgumentException("Session not found: " + id));
    }

    @Transactional
    public ScoreTurnResponse score(UUID sessionId, ScoreRequest request) {
        InterviewSession session =
                sessionRepository
                        .findById(sessionId)
                        .orElseThrow(() -> new IllegalArgumentException("Session not found"));

        Map<String, String> ai =
                temporalService.scoreTurn(
                        sessionId,
                        request.intervieweeText(),
                        request.questionContext(),
                        request.useWebSearch());

        ScoreTurn turn = new ScoreTurn();
        turn.setSession(session);
        turn.setQuestionContext(request.questionContext());
        turn.setIntervieweeText(request.intervieweeText());
        turn.setGrade(ai.get("grade"));
        turn.setRationale(ai.get("rationale"));
        turn = turnRepository.save(turn);

        return toTurnResponse(turn);
    }

    public List<ScoreTurnResponse> listTurns(UUID sessionId) {
        return turnRepository.findBySessionIdOrderByScoredAtAsc(sessionId).stream()
                .map(this::toTurnResponse)
                .toList();
    }

    private SessionResponse toSessionResponse(InterviewSession s) {
        return new SessionResponse(
                s.getId(),
                s.getInterviewerName(),
                s.getCandidateName(),
                s.getRoleTitle(),
                s.getCreatedAt());
    }

    private ScoreTurnResponse toTurnResponse(ScoreTurn t) {
        return new ScoreTurnResponse(
                t.getId(),
                t.getQuestionContext(),
                t.getIntervieweeText(),
                t.getGrade(),
                t.getRationale(),
                t.getScoredAt());
    }
}

File: backend/src/hr_interview_backend/services/interview_service.py
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.orm import Session

from hr_interview_backend.models import InterviewSession, ScoreTurn
from hr_interview_backend.schemas import (
    CreateSessionRequest,
    ScoreRequest,
    ScoreTurnResponse,
    SessionResponse,
)
from hr_interview_backend import temporal_service


def create_session(db: Session, request: CreateSessionRequest) -> SessionResponse:
    session = InterviewSession(
        interviewer_name=request.interviewer_name,
        candidate_name=request.candidate_name,
        role_title=request.role_title,
    )
    db.add(session)
    db.commit()
    db.refresh(session)
    return SessionResponse.model_validate(session)


def get_session(db: Session, session_id: UUID) -> SessionResponse:
    session = db.get(InterviewSession, session_id)
    if session is None:
        raise ValueError(f"Session not found: {session_id}")
    return SessionResponse.model_validate(session)


async def score(db: Session, session_id: UUID, request: ScoreRequest) -> ScoreTurnResponse:
    session = db.get(InterviewSession, session_id)
    if session is None:
        raise ValueError("Session not found")

    ai = await temporal_service.score_turn(
        session_id,
        request.interviewee_text,
        request.question_context,
        request.use_web_search,
    )

    turn = ScoreTurn(
        session_id=session_id,
        question_context=request.question_context,
        interviewee_text=request.interviewee_text,
        grade=ai["grade"],
        rationale=ai["rationale"],
    )
    db.add(turn)
    db.commit()
    db.refresh(turn)
    return ScoreTurnResponse.model_validate(turn)


def list_turns(db: Session, session_id: UUID) -> list[ScoreTurnResponse]:
    stmt = (
        select(ScoreTurn)
        .where(ScoreTurn.session_id == session_id)
        .order_by(ScoreTurn.scored_at.asc())
    )
    turns = db.scalars(stmt).all()
    return [ScoreTurnResponse.model_validate(t) for t in turns]
